src/main/java/net/fabricmc/fabric/container/ContainerHelper.java

```diff
@@ -1,80 +1,72 @@
-		}
-		return null;
-	};
+	private static final Identifier OPEN_CONTAINER = new Identifier("fabric", "open_container");
+	private static final Map<Identifier, ContainerFactory<Container>> FACTORIES = new HashMap<>();
 
-	/**
-	 *
-	 * @param identifier
-	 * @param containerFunction
-	 */
-	public static void registerContainerHandler(Identifier identifier, BiFunction<PlayerEntity, PacketByteBuf, Container> containerFunction){
-		containerMap.put(identifier, containerFunction);
-	}
-
-	public static void registerBlockContainerHandler(Identifier identifier, BiFunction<PlayerEntity, BlockPos, Container> containerFunction){
-		registerContainerHandler(identifier, (playerEntity, packetByteBuf) -> {
-			BlockPos pos = packetByteBuf.readBlockPos();
-			return containerFunction.apply(playerEntity, pos);
-		});
+	public static void registerFactory(Identifier identifier, ContainerFactory<Container> context) {
+		if (FACTORIES.containsKey(identifier)) {
+			throw new RuntimeException("A factory has already been registered as " + identifier.toString());
+		}
+		FACTORIES.put(identifier, context);
 	}
 
 	/**
-	 *
 	 * Sends a pack to the client to open the gui, and opens the container on the server side
 	 *
 	 * @param identifier the identifier that you registered your gui and container handler with
-	 * @param byteBufConsumer a {@link PacketByteBuf} that you can write your own data to
-	 * @param playerEntity the player that the gui should be opened on
+	 * @param writer a {@link PacketByteBuf} that you can write your own data to
+	 * @param player the player that the gui should be opened on
 	 */
-	public static void openGui(Identifier identifier, Consumer<PacketByteBuf> byteBufConsumer, ServerPlayerEntity playerEntity) {
+	public static void openContainer(Identifier identifier, Consumer<PacketByteBuf> writer, ServerPlayerEntity player) {
+		int syncId = 120; //Write a container sync id, does this need to change?
 		PacketByteBuf buf = new PacketByteBuf(Unpooled.buffer());
 		buf.writeString(identifier.toString());
-		byteBufConsumer.accept(buf);
-		playerEntity.networkHandler.sendPacket(new CustomPayloadClientPacket(OPEN_CONTAINER, buf));
-
-		playerEntity.container = containerMap.get(identifier).apply(playerEntity, buf);
-		playerEntity.container.addListener(playerEntity);
-	}
+		buf.writeInt(syncId);
+		writer.accept(buf);
+		player.networkHandler.sendPacket(new CustomPayloadClientPacket(OPEN_CONTAINER, buf));
 
-	/**
-	 *  Opens a gui for a tile that implements FabricContainerProvider
-	 * @param containerProvider the container provider
-	 * @param pos The block pos of the container provider
-	 * @param playerEntity The player that the gui will be opened on
-	 */
-	public static void openGui(FabricContainerProvider containerProvider, BlockPos pos, ServerPlayerEntity playerEntity) {
-		openGui(containerProvider.getContainerIdentifier(), packetByteBuf -> packetByteBuf.writeBlockPos(pos), playerEntity);
+		if (!FACTORIES.containsKey(identifier)) {
```

Comment:
No, don't do this. Just `FACTORIES.get(identifier)` and check if it is `null` first.